lib/stringtheory/src/lib.rs

```diff
@@ -1,58 +1,478 @@
+union Inner {
+    empty: EmptyUnion,
+    owned: OwnedUnion,
+    _static: StaticUnion,
+    interned: ManuallyDrop<InternedUnion>,
+    inlined: InlinedUnion,
+    discriminant: DiscriminantUnion,
+}
+
+impl Inner {
+    const fn empty() -> Self {
+        Self {
+            empty: EmptyUnion::new(),
+        }
+    }
+
+    fn owned(value: String) -> Self {
+        match value.capacity() {
+            // The string we got is empty.
+            0 => Self::empty(),
+            cap => {
+                let mut value = value.into_bytes();
+
+                let ptr = value.as_mut_ptr();
+                let len = value.len();
+
+                // We're taking ownership of the underlying string allocation so we can't let it drop.
+                std::mem::forget(value);
+
+                Self {
+                    owned: OwnedUnion {
+                        ptr,
+                        len,
+                        cap: tag_cap(cap),
+                    },
+                }
+            }
+        }
+    }
+
+    fn static_str(value: &'static str) -> Self {
+        match value.len() {
+            0 => Self::empty(),
+            len => Self {
+                _static: StaticUnion {
+                    ptr: value.as_bytes().as_ptr() as *mut _,
+                    len,
+                    _cap: Zero::Zero,
+                },
+            },
+        }
+    }
+
+    fn interned(value: InternedString) -> Self {
+        match value.len() {
+            0 => Self::empty(),
+            _len => Self {
+                interned: ManuallyDrop::new(InternedUnion {
+                    state: value,
+                    _len: Zero::Zero,
+                    _cap: Zero::Zero,
+                }),
+            },
+        }
+    }
+
+    fn try_inlined(value: &str) -> Option<Self> {
+        match value.len() {
+            0 => Some(Self::empty()),
+            len => {
+                if len > INLINED_STR_MAX_LEN {
+                    return None;
+                }
+
+                let mut data = [0; INLINED_STR_DATA_BUF_LEN];
+
+                // SAFETY: We know it fits because we just checked that the string length is 23 or less.
+                data[INLINED_STR_MAX_LEN] = len as u8;
+
+                let buf = value.as_bytes();
+                data[0..len].copy_from_slice(buf);
+
+                Some(Self {
+                    inlined: InlinedUnion { data },
+                })
+            }
+        }
+    }
+
+    #[inline]
+    fn as_str(&self) -> &str {
+        let union_type = unsafe { self.discriminant.get_union_type() };
+        match union_type {
+            UnionType::Empty => "",
+            UnionType::Owned => {
+                let owned = unsafe { &self.owned };
+                owned.as_str()
+            }
+            UnionType::Static => {
+                let _static = unsafe { &self._static };
```

Comment:
I finally realized that you're using `_static` because `static` is a keyword. I kept focusing on `_foo` being the way to silence the `dead_code` warning. I guess I'd recommend using `static_` or `r#static` instead?

Reply:
Yeah, fair point. Definitely not idiomatic to use a leading whitespace for an actually-used variable.

Reply:
Fixed in c483845 to be `static_`.